executor/load_data.go

```diff
@@ -113,17 +120,142 @@ type LoadDataInfo struct {
+	e.QuitProcess = make(chan struct{})
+}
+
+// ForceQuitCommit let commit quit directly
+func (e *LoadDataInfo) ForceQuitCommit() {
+	close(e.QuitCommit)
+}
+
+// ForceQuitProcess let process quit directly
+func (e *LoadDataInfo) ForceQuitProcess() {
+	close(e.QuitProcess)
+}
+
+// MakeCommitTask produce commit task with data in LoadDataInfo.rows LoadDataInfo.curBatchCnt
+func (e *LoadDataInfo) MakeCommitTask() CommitTask {
+	return CommitTask{e.curBatchCnt, e.rows}
+}
+
+// EnqOneTask feed one batch commit task to commit work
+func (e *LoadDataInfo) EnqOneTask(ctx context.Context) error {
+	var err error
+	if e.curBatchCnt > 0 {
+		e.commitTaskQueue <- e.MakeCommitTask()
+		logutil.BgLogger().Debug("one task enqueued ", zap.Int("current queue len", len(e.commitTaskQueue)))
+		// reset rows buffer, will reallocate buffer but NOT reuse
+		e.SetMaxRowsInBatch(e.maxRowsInBatch)
+	}
+	return err
+}
+
+// CommitOneTask insert Data from LoadDataInfo.rows, then make commit and refresh txn
+func (e *LoadDataInfo) CommitOneTask(ctx context.Context, task CommitTask, resetBuf bool) error {
+	var err error
+	err = e.CheckAndInsertOneBatch(task.rows, task.cnt)
+	if err != nil {
+		logutil.BgLogger().Error("commit error CheckAndInsert", zap.Error(err))
+		return err
+	}
+	if err = e.Ctx.StmtCommit(); err != nil {
+		logutil.BgLogger().Error("commit error commit", zap.Error(err))
+		return err
+	}
+	// Make sure that there are no retries when committing.
+	if err = e.Ctx.RefreshTxnCtx(ctx); err != nil {
+		logutil.BgLogger().Error("commit error refresh", zap.Error(err))
+		return err
+	}
+	// this only set in sequential mode, e.rows buffer will be reused in sequential mode
+	if resetBuf {
+		e.curBatchCnt = 0
+	}
+	return err
+}
+
+// CommitWork commit batch sequentially
+func (e *LoadDataInfo) CommitWork(ctx context.Context) error {
+	var err error
+	defer func() {
+		r := recover()
+		if r != nil {
+			buf := make([]byte, 4096)
+			stackSize := runtime.Stack(buf, false)
+			buf = buf[:stackSize]
+			logutil.Logger(ctx).Error("CommitWork panicked", zap.String("stack", string(buf)))
+		}
+		if err != nil || r != nil {
+			e.ForceQuitProcess()
+		}
+	}()
+	var tasks uint64 = 0
+	var end = false
+	for !end {
+		select {
+		case commitTask, ok := <-e.commitTaskQueue:
+			if ok {
+				err = e.CommitOneTask(ctx, commitTask, false)
+				if err != nil {
+					break
+				}
+				tasks++
+				logutil.BgLogger().Debug("commit one task finished",
+					zap.Uint64("finished tasks count", tasks),
+					zap.Int("pending tasks count", len(e.commitTaskQueue)))
+			} else {
+				end = true
+				logutil.BgLogger().Info("commit work all finished",
+					zap.Uint64("total processed", tasks),
+					zap.Uint64("batch size", e.maxRowsInBatch))
+				break
+			}
+		case <-e.QuitCommit:
+			err = errors.New("commit forced to quit")
+			logutil.Logger(ctx).Error("commit forced to quit, possible preparation failed")
+			break
+		}
+		if err != nil {
+			break
+		}
+	}
+	return err
```

Comment:
maybe we also need think about one corner case:

we are looking 10000 line data, in 100 line's commit failed in consumer worker and return, producer goroutine will keep working and find it can not enqueue data any more and block in that, then no chance to make producer goroutine to exit?

Reply:
`CommitWork` defer call, if CommitWork got errors or panicked, will use `e.ForceQuitProcess()` to inform producing routine to quit. If `CommitWork` quit without error, that means producer normally close task queue already. 
@lysu  PTAL

Reply:
@cfzjywxk  ok, sorry I miss that yesterday...

and maybe we should better check `commitTaskQueue` and `QuitProcess` at the same time? maybe there are some situation goroutine blocked on `commitTaskQueue` and no chance to receive `QuitProcess`?

Reply:
there is such risk,  commit quit and TaskQueue is full. Processing cant enqueue and has not chance to quit,  many thx~ @lysu

Reply:
@lysu  PTAL